File: crates/r2image/tests/data/relocatable.py
#!/usr/bin/env python3
"""Write the relocatable-object fixture: two unplaced sections and two symbols.

A relocatable object states no addresses. Every section carries `sh_addr = 0`,
which means unplaced rather than "at zero", and a symbol's `st_value` is an
offset into its section. The reader has to place the sections itself, and this
fixture is the smallest file that says so: one code section, one data section,
a `STT_FUNC` symbol in the first and a `STT_NOTYPE` label in the second.
"""

import struct

TEXT = bytes([0x48, 0x31, 0xC0, 0xC3, 0x90, 0x90, 0x90, 0x90])  # xor rax, rax; ret
DATA = bytes([0x2A, 0x00, 0x00, 0x00])

names = b"\0.text\0.data\0.symtab\0.strtab\0.shstrtab\0"
offset = {
    name: names.index(b"\0" + name.encode() + b"\0") + 1
    for name in (".text", ".data", ".symtab", ".strtab", ".shstrtab")
}

strtab = b"\0f\0label\0"
symbols = [
    (0, 0, 0, 0, 0),  # the null symbol
    (strtab.index(b"\0f\0") + 1, (1 << 4) | 2, 0, 1, 0),  # LOCAL FUNC .text +0
    (strtab.index(b"\0label\0") + 1, 0, 0, 2, 0),  # LOCAL NOTYPE .data +0
]
symtab = b"".join(
    struct.pack("<IBBHQQ", name, info, 0, shndx, value, 0)
    for name, info, _other, shndx, value in symbols
)

text_off = 0x40
data_off = text_off + len(TEXT)
symtab_off = data_off + len(DATA)
strtab_off = symtab_off + len(symtab)
shstrtab_off = strtab_off + len(strtab)
shoff = shstrtab_off + len(names)

def header(name, kind, flags, off, size, link, info, align, entsize):
    return struct.pack(
        "<IIQQQQIIQQ", name, kind, flags, 0, off, size, link, info, align, entsize
    )

sections = b"".join(
    [
        header(0, 0, 0, 0, 0, 0, 0, 0, 0),
        header(offset[".text"], 1, 0x6, text_off, len(TEXT), 0, 0, 1, 0),
        header(offset[".data"], 1, 0x3, data_off, len(DATA), 0, 0, 1, 0),
        header(offset[".symtab"], 2, 0, symtab_off, len(symtab), 4, len(symbols), 8, 24),
        header(offset[".strtab"], 3, 0, strtab_off, len(strtab), 0, 0, 1, 0),
        header(offset[".shstrtab"], 3, 0, shstrtab_off, len(names), 0, 0, 1, 0),
    ]
)

elf = bytearray(b"\x7fELF\x02\x01\x01\x00" + b"\0" * 8)
elf += struct.pack("<HHIQQQIHHHHHH", 1, 0x3E, 1, 0, 0, shoff, 0, 64, 0, 0, 64, 6, 5)
elf += b"\0" * (text_off - len(elf))
elf += TEXT + DATA + symtab + strtab + names + sections

import pathlib

pathlib.Path(__file__).with_name("relocatable.elf").write_bytes(bytes(elf))
print("wrote relocatable.elf", len(elf), "bytes")

File: crates/r2image/tests/relocatable.rs
//! A relocatable object states no addresses, so the reader places its sections.
//!
//! Every section of an ELF object carries `sh_addr = 0`, which means unplaced
//! rather than "at zero". Reading that as an address put every section at the
//! same place, so a read answered from whichever came first and disassembling
//! such a file produced nothing. The fixture is written by `data/relocatable.py`.

const BASE: u64 = 0x0800_0000;

fn image() -> r2image::Image {
    let bytes = include_bytes!("data/relocatable.elf").to_vec();
    r2image::Image::parse(bytes).expect("a relocatable elf fixture")
}

#[test]
fn each_unplaced_section_is_placed_at_its_own_file_offset() {
    let image = image();
    let placed = |name: &str| {
        image
            .sections()
            .iter()
            .find(|section| section.name == name)
            .map(|section| section.vaddr)
    };
    assert_eq!(placed(".text"), Some(BASE + 0x40));
    assert_eq!(placed(".data"), Some(BASE + 0x48));
}

#[test]
fn the_placed_bytes_are_the_bytes_of_that_section() {
    let image = image();
    let text = image.read(BASE + 0x40, 4).expect("the code is mapped");
    assert_eq!(&*text, &[0x48, 0x31, 0xc0, 0xc3]);
    let data = image.read(BASE + 0x48, 4).expect("the data is mapped");
    assert_eq!(&*data, &42u32.to_le_bytes());
}

/// A symbol's `st_value` is an offset into its section, not an address, and an
/// untyped symbol is defined as surely as a typed one: every NASM label is
/// `STT_NOTYPE`, and a binary made of them listed no symbols at all.
#[test]
fn a_symbol_moves_with_its_section_whatever_its_type() {
    let image = image();
    let at = |name: &str| {
        image
            .symbols()
            .iter()
            .find(|symbol| symbol.name == name)
            .map(|symbol| (symbol.vaddr, symbol.defined))
    };
    assert_eq!(at("f"), Some((BASE + 0x40, true)));
    assert_eq!(at("label"), Some((BASE + 0x48, true)));
}
